sdk/storage/Azure.Storage.Blobs/src/BlockBlobWriteStream.cs

```diff
@@ -0,0 +1,111 @@
+    {
+        private readonly BlockBlobClient _blockBlobClient;
+        private readonly BlobRequestConditions _conditions;
+        private readonly List<string> _blockIds;
+
+        public BlockBlobWriteStream(
+            BlockBlobClient blockBlobClient,
+            int bufferSize,
+            long position,
+            BlobRequestConditions conditions,
+            IProgress<long> progressHandler) : base(
+                position,
+                bufferSize,
+                progressHandler)
+        {
+            ValidateBufferSize(bufferSize);
+            _blockBlobClient = blockBlobClient;
+            _conditions = conditions;
+            _blockIds = new List<string>();
+        }
+
+        protected override async Task AppendInternal(bool async, CancellationToken cancellationToken)
+        {
+            if (_buffer.Length > 0)
+            {
+                _buffer.Position = 0;
+
+                string blockId = GenerateBlockId();
+
+                if (async)
+                {
+                    await _blockBlobClient.StageBlockAsync(
+                        base64BlockId: blockId,
+                        content: _buffer,
+                        conditions: _conditions,
+                        progressHandler: _progressHandler,
+                        cancellationToken: cancellationToken)
+                        .ConfigureAwait(false);
+                }
+                else
+                {
+                    _blockBlobClient.StageBlock(
+                        base64BlockId: blockId,
+                        content: _buffer,
+                        conditions: _conditions,
+                        progressHandler: _progressHandler,
+                        cancellationToken: cancellationToken);
+                }
+
+                _blockIds.Add(blockId);
+                _buffer.SetLength(0);
+            }
+        }
+
+        protected override async Task FlushInternal(bool async, CancellationToken cancellationToken)
+        {
+            await AppendInternal(async, cancellationToken).ConfigureAwait(false);
+
+            if (async)
+            {
+                await _blockBlobClient.CommitBlockListAsync(
+                    base64BlockIds: _blockIds,
+                    conditions: _conditions,
+                    cancellationToken: cancellationToken)
+                    .ConfigureAwait(false);
+            }
+            else
+            {
+                _blockBlobClient.CommitBlockList(
+                    base64BlockIds: _blockIds,
+                    conditions: _conditions,
+                    cancellationToken: cancellationToken);
+            }
+        }
+
+        protected override void ValidateBufferSize(int bufferSize)
+        {
+            if (bufferSize < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be >= 1");
+            }
+
+            if (bufferSize > 100 * Constants.MB)
```

Comment:
Why 100MB?

Reply:
Put block is limited to 100 MB.  https://docs.microsoft.com/en-us/rest/api/storageservices/put-block#request-headers

Reply:
It's going to 4000 pretty soon.

Reply:
I'm going to need to use partitioned uploader's buffer solution as you suggested below then.

Reply:
Yeah, I think we should do that and ship this feature right after stg73 so we don't need to rework this or break (`long`s).

Reply:
I'll raise this when we need to, and bufferSize has been switched to a long.